EOS-13276: DSAL: Unmap of truncated object needs to be implemented.

gshipra opened the pull request:
Problem Statement :
EOS-13276: DSAL: Unmap of truncated object needs to be implemented.
Problem Description:
Need to add unmap functionality.
Solution Overview:
Have added code to implement unmap functionality.
Unit Test Cases:
1.UT passes.
2. Newly added UT passes.
3. nfs_posix test passes.
4. Cthon basic, general, special, lock tests passes.

Signed-off-by: Shipra <shipra.gupta@seagate.com>

ivan-alekhin commented on src/include/dstore.h:
```diff
@@ -182,4 +184,11 @@ int dstore_pwrite(struct dstore_obj *obj, off_t offset, size_t count,
+ *  @param[in] vec - vector filled with offset and count.
+ *  @return 0 or -errno.
+ */
+int dstore_ptrunc(struct dstore_obj *obj, struct dstore_extent_vec *vec);
```
What do you say about the following interface?

```c
enum dstore_obj_fallocate_mode {
  /* De-allocate storage space allocated by by Alloc or Write operations */
  DSTORE_OBJ_FALLOC_FREE,
};

/* Allocates or de-allocates synchronously the range specified by offset and len.
 * When the range is not aligned with the block size of the underlying storage,
 * it fills the leftovers with zeros.
 * TODO: Allocate mode is not implemented yet.
 * TODO: Unaligned ranges are not handled yet.
 * @param mode Operation to be perform (see enum ::dstore_obj_fallocate_mode).
 */
int dstore_fallocate(struct dstore_obj *obj, uint64_t mode, off_t offset, off_t len);
```

We don't need the ability to free a group of extents here: truncate() changes the file size continuously -- there are no gaps, so that it is enough to have only one extent.

As usually we have a choice here: either provide an interface that reassembles m0 client API (with operations, io/index vectors and callbacks) or an interface that reassembles simple POSIX-like API (single range, synchronous). Since the previous operations (pwrite, pread) are quite close to their POSIX counterparts, let's see if we can make our Alloc/Free to be similar to its POSIX equivalent (fallocate).

gshipra replied:
I have done new posix formta implementation. I want to keep api separate for alloc and dealloc.

ivan-alekhin commented on src/dstore/dstore_internal.h:
```diff
@@ -456,6 +465,18 @@ struct dstore_ops {
+	 * This function creates a new unmap operation using the
+	 * given inputs.
+	 */
+	int (*io_trunc_op_init)(struct dstore_obj *dobj,
```
I wonder if we really need to add one more interface function here.

We could try to change `dstore_io_vec` to support extents-only (no IO buffers). If no-IO-buffers io_vec does not hold the invariant, then `dstore_io_vec.flags` might help us with that: we can introduce a flag "NO_IO_DATA" and when `dstore_io_vec.flags & NO_IO_DATA` is set then the invariant check will be ignoring data buffers and theirs sizes.

What do you think about that?

gshipra replied:
Thanks for guiding me. I was feeling confused at this particular section. I have used the flags as suggested. Please do the review and let me know if I have implemented correctly.

ivan-alekhin reviewed (commented):
The implementation part looks fine to me but I have questions/suggestions about the interfaces.

ivan-alekhin commented on src/dstore/dstore_internal.h:
```diff
@@ -28,6 +28,9 @@
 #include "dstore.h" /* import public data types */
 
+#define IO_DATA  0x001
+#define IO_NO_DATA    0x002
```
Let's convert it into an enum -- they are easier to be used with a debugger.

Also, I guess it is enough to have only one flag. We cannot have them both set, so probably
it is enough to have only one.

```c
enum dstore_io_vec_flags {
    DSTORE_IVF_NO_IO_DATA = 1 << 1,
}

static bool dstore_io_vec_flags_has_data(uint64_t flags)
{
  return (flags & DSTORE_IVF_NO_DATA) == 0;
}
```

Then you can replace `io_vec->flags & IO_NO_DATA` with `!dstore_io_vec_flags_has_data(io_vec->flags)` and `io_vec->flags & IO_DATA` with `dstore_io_vec_flags_has_data(io_vec->flags)`. We don't have to use this particular name and we even don't have to introduce this function (it can be a define or just bare expression). The main idea here is to avoid ambiguity whenever it is possible.

If you would like to keep both flags then please add a condition to the invariant for the vector:

```c
bool consistent_flags = !(flags & IO_NO_DATA) == !!(flags & IO_DATA);
```

gshipra replied:
I have changed the name of flag and added function told by you. There is only one flag now. But I could not make it enum.

ivan-alekhin commented on src/dstore/dstore_internal.h:
```diff
@@ -211,6 +214,10 @@ bool dstore_io_vec_invariant(const struct dstore_io_vec *io_vec)
 	 * NOTE: This condition may be changed in future if
 	 * we need to implement Alloc/Free operations.
 	 */
+	if (io_vec->flags & IO_NO_DATA) {
```
Let's move it part to the very beginning of this function and explain it:

```c
/* When NO_DATA flag is set, the vector invariant cannot be violated:
 *   the embedded buf has only offset, size -- they are always consistent.
 *   consistency of the extents (when embedded buf is not used) is not checked
 *     anyway, and it is not assumed.
 * Therefore, it is safe to assume that when this flag is set, we have
 * a valid io_vec object.
 */
```

gshipra replied:
Done.

ivan-alekhin commented on src/dstore/dstore_base.c:
```diff
@@ -799,3 +757,125 @@ int dstore_pread(struct dstore_obj *obj, off_t offset, size_t count,
+                  "offset = %lu count = %lu rc = %d",
+                  OBJ_ID_P(dstore_obj_id(obj)), obj, offset, count, rc);
+
+
```
extra newline. Let's delete it.

gshipra replied:
Done.

ivan-alekhin reviewed (commented):
Thanks for the the fixes to the previous round of review. I posted a few minor suggestions. I guess the next round will be final from my side -- overall the patch looks good.

Anilsindhi commented on src/dstore/dstore_base.c:
```diff
@@ -31,6 +31,8 @@
 #include "dstore_bufvec.h" /* data buffers and vectors */
 #include "operation.h"
 
+#define DSAL_MAX_IO_SIZE (1024*1024)
```
This macro is already defined at line number 125, the reason for defining it was we were not able to do write of more than 1MB but it is not the case with the truncate, you can do a single truncate operation upto 20MB as per old code implementation. Please refer following code snippet from old code.

```
/* NSAL_TUNEABLE:
 * Size of data block that can be definitely UNMAP-ed by Motr
 * without generating errors at RPC layer or getting stuck somewhere
 * in the state machine.
 */
/* Default value: 5120 4K pages or 20 1MB pages or 20MB of data */
static const uint64_t m0_cortxfs_trunc_data_per_request = 20 * (1 << 20);

/** Submits UNMAP requests to Clvois and waits until the data blocks
 * are actually unmapped.
 * NOTE: Motr is not able to handle large extents in the truncate operations,
 * so that we are sending only small portions of extents per request.

```

So you can do two things here.

1. Remove macro define at line number 125 as it is no longer needed and TODO mentioned along with that.
2. Rename this macro specific to truncate operation something like DSAL_MAX_TRUNC_OP_SIZE (20 * 1024 * 1024)  with this you will be submitting lesser truncate request to motr. I don't see any issue with keeping it 20MB as it was already working in old code.

gshipra replied:
Done.

Anilsindhi commented on src/dstore/dstore_base.c:
```diff
@@ -799,3 +757,125 @@ int dstore_pread(struct dstore_obj *obj, off_t offset, size_t count,
+		write_count = bsize - write_count;
+
+		if (count < write_count) {
+
```
Extra new line, let's delete it.

gshipra replied:
Done.

Anilsindhi commented on src/dstore/dstore_base.c:
```diff
@@ -799,3 +757,125 @@ int dstore_pread(struct dstore_obj *obj, off_t offset, size_t count,
+	ndata_per_req = DSAL_MAX_IO_SIZE;	
+	nr_request = count / ndata_per_req;
+	tail_size = count - (nr_request * ndata_per_req);
+	nblk_per_req = ndata_per_req / bsize;
```
Remove this as it is not used anywhere.

gshipra replied:
Done.

Anilsindhi commented on src/dstore/dstore_base.c:
```diff
@@ -799,3 +757,125 @@ int dstore_pread(struct dstore_obj *obj, off_t offset, size_t count,
+		offset = offset + write_count;
+	}
+
+	ndata_per_req = DSAL_MAX_IO_SIZE;	
```
As mentioned in earlier comment division should happen with 20MB so less number of truncate request to Motr.

gshipra replied:
Done.

Anilsindhi commented on src/dstore/dstore_base.c:
```diff
@@ -799,3 +757,125 @@ int dstore_pread(struct dstore_obj *obj, off_t offset, size_t count,
+	dassert(obj);
+
+	char *tmp_buf = NULL;
+	struct dstore_io_op *wop = NULL;
```
Would be good if you rename this variable to some proper name as we aren't using it for doing write operation(wop) we are using it for freeing blocks.

gshipra replied:
Done.

Anilsindhi commented on src/dstore/dstore_base.c:
```diff
@@ -799,3 +757,125 @@ int dstore_pread(struct dstore_obj *obj, off_t offset, size_t count,
+
+	log_trace("dstore_deallocate:(" OBJ_ID_F " <=> %p )"
+                  "offset = %lu count = %lu rc = %d",
+                  OBJ_ID_P(dstore_obj_id(obj)), obj, offset, count, rc);
```
If you see line number 862 and 846, you are modifying `offset` and then logging it. So ultimately you are not logging the original input `offset` but modified one and it may mislead. Let's say you came here with `rc = 0` it means all operation completed successfully then at this place your `count` and `offset` will be same and we will be wondering how did we reached here if we have `count == offset`.

gshipra replied:
Done.

Anilsindhi commented on src/dstore/dstore_base.c:
```diff
@@ -799,3 +757,125 @@ int dstore_pread(struct dstore_obj *obj, off_t offset, size_t count,
 	return rc;
 }
+
+int dstore_dealloc_op(struct dstore_obj *obj, struct dstore_io_vec *vec,
```
Make this API internal... `static int ...` same for `dstore_deallocate` too.

gshipra replied:
Done.

Anilsindhi commented on src/dstore/dstore_base.c:
```diff
@@ -799,3 +757,125 @@ int dstore_pread(struct dstore_obj *obj, off_t offset, size_t count,
+	rc = dstore_io_op_init_and_submit(obj, vec, out,
+					  DSTORE_IO_OP_FREE);
+
+	log_debug("dstore_dealloc_op (" OBJ_ID_F " <=> %p, "
```
No need to write function name here, as per your recent change in other MR I understood line number and function name will be take care by `log_debug`. I know that we have done the same at lot's of place and we have to remove that but let's not add such new entries.

Please remove it whichever API you have modified.

gshipra replied:
Done.

Anilsindhi commented on src/dstore/dstore_base.c:
```diff
@@ -799,3 +757,125 @@ int dstore_pread(struct dstore_obj *obj, off_t offset, size_t count,
+	}
+
+	if (tail_size) {
+
```
Question: @ivan-alekhin  and @gshipra , Do we really need this right-most unaligned part? I believe we will never be un-aligned in right-most part because while writing this data we would have taken care and written always in aligned boundary.

NOTE: In old code also we were not doing `MC_OC_FREE` for right most part.

ivan-alekhin replied:
that's a good question Anil. 

In my original code the idea was to round up the number of blocks, so that the right part (if is not aligned) will be de-allocated:

```c
	/* ajust to the very first byte of the nearest (from right) page */
	aligned_off = m0_round_up(offset, bsize);
	/* cut out the left unaligned part from the whole len,
	 * round it up and then count the amount of blocks */
	nblocks = m0_round_up((count + offset) - aligned_off, bsize) / bsize;
// ...
	rc = m0_file_unmap_aligned(fid, nblocks, aligned_off, bsize);

// and then inside unmap_aligned:
	ndata_per_req = lower(trunc_data_per_request, bsize);
	nblk_per_req = ndata_per_req / bsize;
	nrequests = nblocks / nblk_per_req;
	tail_size = (nblocks * bsize) - (nrequests * ndata_per_req);
```

So that, `tail_size` is juts the aligned leftover from the right side. It should be aligned with the block size but should not be aligned with the request size. The following invariant should be held for it: `tail_size % block_size == 0 && tail_size < ndata_per_req`.

To conclude, I guess we are handling here unaligned right side here in a wrong way: if `offset+count % bs != 0`  then tail_size also won't be aligned (`tail_size % bs != 0`). It means M0 client will tell us that we using the wrong input.

@gshipra , what do you think about moving the portion of code that handles aligned FREE operation into its own function (as it was in the original code)? it will help us to analyze the code.

gshipra replied:
@Anilsindhi  - This is not right alignment for data buffer. We are dividing FREE in multiple of 20 MB. Lets sat we have 21 MB to free. Tail size here is this 21%20 = 1 MB.

gshipra replied:
@Anilsindhi  - Consider follwoing scenarios.

3000 bytes are written to a file at offset 0, we will round up write request and actually write 4096 byes. File Size is 3000.

1. Trunc request to truncate file to 0. Now offset is 0 and we need to free 3000 bytes.
We will send count as 3000 to free operation as per current implementation. But motr does not handle this. It panics.
Motr needs aligned count. I figured this out while testing. 
When I tried to free 4096 bytes, operation succeeded.

2. Lets say truncate sets the new file size to 9000. As per current implementation we wont do any thing other than increasing the size.
0-3000 - valid data
3000 - 4095 - 0s
4096-8191 - hole
8192 - 9000 - ?????

gshipra replied:
@Anilsindhi , @ivan-alekhin - please suggest which function is better to do left+right alignment calculataions ? dstore_shrink or dstore_deallocate ?

Anilsindhi replied:
Okay I understood the use case. But just the question on your reply.
Your offset = 0, count = 3000 so,
```
if (offset % bsize != 0) {
// We won't go here as offset is zero so offset % bsize == 0
}

ndata_per_req = 20MB;
nr_request = count / ndata_per_req = 3000/20MB = 0;
tail_size = count - (nr_request * ndata_per_req) = 3000 - (0 *20MB) = 3000; // Not aligned to bsize

for (i = 0; i < nr_request; i++) {
      // We will not come here as `nr_request` is 0
}

// In the following code you are still issuing unaligned unmap request to motr.

	if (tail_size) {
		*(vec.ovec) = offset;
		*(vec.svec) = tail_size; // This is 3000 and not rounded up to bsize

		RC_WRAP_LABEL(rc, out, dstore_dealloc_op, obj, &vec, &dop);

		RC_WRAP_LABEL(rc, out, dstore_io_op_wait, dop);

		dstore_io_op_fini(dop);

		dop = NULL;
	}
```
If my understanding is correct then still the current code is not behaving as expected.

ivan-alekhin replied:
@gshipra ,

> please suggest which function is better to do left+right alignment calculataions ? dstore_shrink or dstore_deallocate ?

dstore_shrink uses dstore_pwrite. dstore_pwrite is able to handle unaligned ranges for WRITE operation. But there are no function that can handle unaligned ranges for FREE operation. In fact, dstore_deallocate will be the function that will handle this. Therefore, you cannot use dstore_shrink as a reference. So, I'd recommend to look at the original code (
m0_file_unmap) -- there are two separate functions. One function handles any aligned range (by grouping blocks into bigger requests), and the other handles alignment (by writing zeros on the left side and freeing full blocks on the right side).

gshipra replied:
@Anilsindhi - I have told in previous comments that two scenarios are discovered via new UT code added and are not handled in code yet. You mentioned in a comment before that we do not need to do right alignment but it seems we need to do right alignment else motr may panic for last free request.

Anilsindhi replied:
@gshipra, Okay I understood now that you are aware of the issue which I highlighted, good that we have written UT along with this change and we caught the issue. Please go through @ivan-alekhin comments in this thread, and look at the old code as per his suggestion and you can something out of that.

gshipra replied:
Added the right alignment code.

Anilsindhi replied:
I am fine with this changes, @ivan-alekhin if you are also fine then feel free to close this comment.

ivan-alekhin replied:
Yes, the recent changes look good.

Anilsindhi reviewed (commented):
Have done a first round of review. Please address the comments which ever applicable.

1. Requesting to add the testing details in PR description like UT is passing, cthon is working and basic I/O should work from NFS client. If all of this testing is not done then need to be done.
2. There aren't any specific test cases developed to verify truncate operation, would be good to add those.

gshipra commented on src/dstore/dstore_base.c:
```diff
@@ -799,3 +757,125 @@ int dstore_pread(struct dstore_obj *obj, off_t offset, size_t count,
+	tail_size = count - (nr_request * ndata_per_req);
+	nblk_per_req = ndata_per_req / bsize;
+
+	vec = calloc(1, sizeof(struct dstore_io_vec));
```
Allocated on stack.

gshipra commented on src/dstore/dstore_internal.h:
```diff
@@ -456,6 +464,7 @@ struct dstore_ops {
 
 	/* This function returns block size */
 	ssize_t (*obj_get_bsize) (dstore_oid_t *oid);
+
```
Done.

Anilsindhi replied:
Still extra line is not removed, Please do it.

gshipra replied:
done.

gshipra commented:
1. I have run UT and made changes to code for UT to pass. Please review the code again.
2. Implemented review comments.
3. Added new tests cases for truncate. 
TO DO: Two more scenarios needs to be handled. UT also needs to test these scenarios.

Anilsindhi commented on src/dstore/dstore_base.c:
```diff
@@ -817,3 +773,117 @@ int dstore_pread(struct dstore_obj *obj, off_t offset, size_t count,
+{
+	int rc = 0;
+
+	dassert(obj);
```
Move this after variable list.

gshipra replied:
done.

Anilsindhi commented on src/dstore/dstore_internal.h:
```diff
@@ -202,15 +204,24 @@ struct dstore_io_vec {
 	struct dstore_io_buf edbuf;
 };
 
+static bool dstore_io_vec_flags_has_data(uint64_t flags) {
```
Make it `inline` if you are adding function definition inside header file,  as per my knowledge generally we add inline function inside header file.

gshipra replied:
done.

Anilsindhi commented on src/include/dstore.h:
```diff
@@ -182,4 +182,5 @@ int dstore_pwrite(struct dstore_obj *obj, off_t offset, size_t count,
  */
 int dstore_pread(struct dstore_obj *obj, off_t offset, size_t count,
 		       size_t bs, char *buf);
+
```
Reason for adding this new line? Apart from this no other change in this file.

gshipra replied:
I added a declaration here. Later removed it and newline remained. Removed new line.

Anilsindhi commented on src/test/ut/dsal_test_io.c:
```diff
@@ -310,6 +310,126 @@ static void test_aligned_unaligned_io(void **state)
+	/* TEST CASE 3 */
+
+	/* Write at offset 4096, count = 3096 bytes.
+	 * old file size if 7192, decrease file size to 3096.
```
`old file size is`

gshipra replied:
done.

Anilsindhi commented on src/test/ut/dsal_test_io.c:
```diff
@@ -310,6 +310,126 @@ static void test_aligned_unaligned_io(void **state)
+	/* Write at offset 4096, count = 3096 bytes.
+	 * old file size if 7192, decrease file size to 3096.
+	 * This will test scenario where old file size is not
+	 * not aligned to block size and new file size is also
```
nitpick:Two time not

gshipra replied:
done.

Anilsindhi reviewed (commented):
Few minor comment, address if applicable.
Requesting do do `cthon` testing and update the details in description as in that truncate path will be validated.

ivan-alekhin commented on src/dstore/dstore_base.c:
```diff
@@ -817,3 +773,125 @@ int dstore_pread(struct dstore_obj *obj, off_t offset, size_t count,
+	dassert(obj);
+
+	bsize = dstore_get_bsize(obj->ds,
+				 (dstore_oid_t *)dstore_obj_id(obj));
```
style suggestion: add space there -- `(dstore_oid_t *) dstore_obj_id(obj));`.

gshipra replied:
Done.

ivan-alekhin commented on src/dstore/dstore_base.c:
```diff
@@ -817,3 +773,125 @@ int dstore_pread(struct dstore_obj *obj, off_t offset, size_t count,
+	if (offset % bsize != 0) {
+		/*we need to make deallocate operation left aligned */
+		tmp_buf = calloc(bsize, sizeof(char));
+		uint32_t left_blk_num = offset/bsize;
```
style: spaces around arithmetic operators -- `offset / bsize`

gshipra replied:
done.

ivan-alekhin commented on src/dstore/dstore_base.c:
```diff
@@ -817,3 +773,125 @@ int dstore_pread(struct dstore_obj *obj, off_t offset, size_t count,
+
+	if (offset % bsize != 0) {
+		/*we need to make deallocate operation left aligned */
+		tmp_buf = calloc(bsize, sizeof(char));
```
Let's check the return value of calloc. Even though it is just one single page (https://vorpus.org/blog/why-does-calloc-exist/) and it will be like that its whole life (we write nothing in there, therefore COW will never be turned on), we still have to comply with the rules.

gshipra replied:
Done.

ivan-alekhin commented on src/dstore/dstore_base.c:
```diff
@@ -817,3 +773,125 @@ int dstore_pread(struct dstore_obj *obj, off_t offset, size_t count,
+		/*we need to make deallocate operation left aligned */
+		tmp_buf = calloc(bsize, sizeof(char));
+		uint32_t left_blk_num = offset/bsize;
+		uint32_t write_count = offset - (left_blk_num * bsize);
```
Let's keep variable declarations at the beginning of a function. We are either C99 or C89. It would be better to avoid mixing them.

gshipra replied:
done.

ivan-alekhin commented on src/dstore/dstore_base.c:
```diff
@@ -817,3 +773,125 @@ int dstore_pread(struct dstore_obj *obj, off_t offset, size_t count,
+		offset = offset + write_count;
+	}
+
+	ndata_per_req = (DSAL_MAX_DEALLOC_OP_SIZE/bsize)*bsize;	
```
style: please add spaces around arithmetic operators.

gshipra replied:
done.

ivan-alekhin commented on src/dstore/dstore_base.c:
```diff
@@ -817,3 +773,125 @@ int dstore_pread(struct dstore_obj *obj, off_t offset, size_t count,
+	tail_size = count - (nr_request * ndata_per_req);
+
+	/* we do not need io buffer here, we just need to send extents info */
+	memset(&vec, 0, sizeof(struct dstore_io_vec));
```
We can initialize it with zeros in the beginning: `struct dstore_io_vec vec = { .data = NULL };`.

gshipra replied:
done.

gshipra replied:
I am not able to do this, compile complains error: unknown field ‘data’ specified in initializer.
I do not wish to add each individual element init state there.

ivan-alekhin replied:
sorry, I used the wrong name. It can be dbufs or any other field.

> I do not wish to add each individual element init state there.

It is not required. The C standard guarantees that the rest of the fields will be initialized with zeros:
```c
struct foo { int a; int *b; }
...
void f() {
 struct foo foo = { .a = 0 };
 assert(foo.b == NULL);
}
```

Anyway, if you believe memset looks ok here then let's keep using memset -- it is totally fine, we don't have a special coding style rule for that.

ivan-alekhin commented on src/dstore/dstore_base.c:
```diff
@@ -817,3 +773,125 @@ int dstore_pread(struct dstore_obj *obj, off_t offset, size_t count,
+
+	for (i = 0; i < nr_request; i++) {
+
+		*(vec.ovec) = offset;
```
I guess `vec.ovec[0] = offset` would look more naturally: we are operating on a vector that has only one item. But it is still a vector.

ivan-alekhin reviewed (approved):
From design and functional perspective the changes look good to me.

I posted some minor suggestions and comments.

Anilsindhi reviewed (approved):
LGTM, +1.

gshipra commented:
CTHON result:
Congratulations, you passed the basic tests!
General tests complete
Special tests complete
Congratulations, you passed the locking tests!
Congratulations, you passed the locking tests!

UT passes along with new written UT for truncate.

sachinpunadikar commented on src/dstore/dstore_base.c:
```diff
@@ -289,7 +244,8 @@ static int dstore_io_op_init_and_submit(struct dstore_obj *obj,
 	dassert(dstore_io_vec_invariant(bvec));
 	/* Only WRITE/READ is supported so far */
```
Please update the comment for free/truncate/unmap operation

gshipra replied:
Done. Code Merged. Rand UT and CTHON again.

sachinpunadikar reviewed (approved):
+1
A minor comment to update a comment
